### DISCUSSION TITLE
COBOL folding rules

### DISCUSSION THREAD

**Discussion by FALLAI-Denis** [zokugun/vscode-explicit-folding#60] [Show and tell]
Hi,

I offer you a folding solution for the COBOL language based on the structuring of the language in different levels of hierarchies: Division, Section, Paragraph.

This solution also manages the folding of comments, as well as a folding on the page eject, (inheritance of the time when we used punch cards and Caroll tape listings).

As an enrichment of this solution, the “END-xxxx” termination instructions could be taken into account (example: IF / END-IF).

```
    "[cobol]": {
        "files.autoGuessEncoding": true,
        "editor.rulers": [0,{"column":6,"color":"#603020"},7,11,{"column":72,"color": "#603020"},{"column": 80,"color": "#ff0000"}],
        /* Requirement for Explicit Folding */
        "editor.folding": true,
        "editor.foldingHighlight": true,
        "editor.foldingStrategy": "auto",
        "editor.showFoldingControls": "always",
        /* Explicit Folding */
        "explicitFolding.debug": false,
        "explicitFolding.rules": [
          {// Comments block
           "name": "comment"
          ,"whileRegex": "^.{6}\\*"
          ,"kind": "comment"
          }
          // COBOL hierarchy
         ,{// Division
           "name": "division"
          ,"separatorRegex": "^.{6} {1,4}[A-Za-z0-9\\-_:]+ +(?i:DIVISION)"
          ,"strict": "never"
          ,"nested": [
             {// Section
              "name": "section"
             ,"separatorRegex": "^.{6} {1,4}[A-Za-z0-9\\-_:]+ +(?i:SECTION)"
             ,"nested": [
                {// Paragraph
                 "name": "paragraph"
                ,"separatorRegex": "^.{6} {1,4}[A-Za-z0-9\\-_:]+(?! +(?i:SECTION|DIVISION))"
                ,"nested": [
                   {// Page eject
                    "name": "eject"
                   ,"separatorRegex":"^.{6}\\/"
                   }
                 ]
                }
              ]
             }
           ]
          }
         ]
        },
```

Tested with:
- VS Code 1.58.0
- Explicit Folding 0.16.0
- IBM Z Open Editor 1.2.5

**Comment by FALLAI-Denis** [comment]
Illustration of folding according to COBOL structures.

Possible use of keyboard shortcuts:
- Ctrl+K Ctrl+J : unfold all
- Ctrl+K Ctrl+1 : fold at level 1 (Ctrl+K Ctrl+2 fold at level 2, etc.)

![FoldingCobol](https://user-images.githubusercontent.com/63861965/125813466-bb6f5ef6-5e55-4097-b549-3029f83ba97e.gif)

**Comment by daiyam** [comment]
@FALLAI-Denis Here you are: https://marketplace.visualstudio.com/items?itemName=zokugun.cobol-folding
